src/simple-client.toit

```diff
@@ -0,0 +1,472 @@
+  if $qos is 1, then the message is sent, and the broker will acknowledge
+    that it received the message. This function blocks until the acknowledgment
+    is received. Consider using $with-timeout to avoid blocking indefinitely.
+  */
+  publish topic/string payload/io.Data --qos/int=0 --retain/bool=false:
+    packet-id := next-packet-id_
+    payload-bytes := payload is ByteArray ? payload : ByteArray.from payload
+    packet := PublishPacket topic payload-bytes
+        --qos=qos
+        --retain=retain
+        --packet-id=(qos != 0 ? packet-id : null)
+    if qos == 1:
+      send-and-wait-for-ack_ packet --packet-id=packet-id --kind="PubAck"
+    else:
+      send-packet_ packet
+
+  /**
+  Subscribes to the given $topic.
+
+  The $topic might have wildcards and is thus more of a filter than a single topic.
+
+  The $max-qos is the maximum quality of service that the client wants to receive
+    messages with. The broker might send messages with a lower QoS, but never with
+    a higher QoS. If the broker received a message with a higher QoS, it will send
+    the message with the reduced QoS to the client.
+
+  Blocks until the subscription is acknowledged.
+  */
+  subscribe topic/string --max-qos/int=1:
+    packet-id := next-packet-id_
+    topic-qos := [TopicQos topic --max-qos=max-qos]
+    packet := SubscribePacket topic-qos --packet-id=packet-id
+    send-and-wait-for-ack_ packet --packet-id=packet-id --kind="SubAck"
+
+  /**
+  Unsubscribes from the given $topic.
+
+  Blocks until the unsubscription is acknowledged.
+  */
+  unsubscribe topic/string:
+    packet-id := next-packet-id_
+    packet := UnsubscribePacket [topic] --packet-id=packet-id
+    send-and-wait-for-ack_ packet --packet-id=packet-id --kind="UnsubAck"
+
+  /**
+  The number of received packets that are currently in the queue.
+  */
+  received-count -> int:
+    return received-queue_.size
+
+  /**
+  Receives a packet.
+
+  Blocks until a packet is received.
+  Returns null if the client is closed.
+  The $PublishPacket.payload is already read into memory.
+  */
+  receive -> PublishPacket?:
+    e := catch --unwind=(: state_ != STATE-DISCONNECTED_ or state_ == STATE-CLOSED_):
+      connection-monitor_.wait:
+        loock-for-received_ --allow-close --kind="receive" --without-mutex:
+          not received-queue_.is-empty
+      if received-queue_.is-empty:
+        // We are closed.
+        return null
+      return received-queue_.remove-first
+    return null
+
+  receive-publish-packet_ packet/PublishPacket:
+    received-queue_.add packet
+    if packet.qos == 0: return
+    if packet.qos == 1:
+      logger_.debug "sending ack" --tags={"packet-id": packet.packet-id}
+      ack := PubAckPacket --packet-id=packet.packet-id
+      catch --trace: send-packet_ ack
+
+monitor MutexSignal:
+  do [block]:
+    block.call
+
+  wait [block]:
+    await block
+
+  do-wait [--do] [--wait]:
```

Comment:
Maybe add a comment here to explain why this is different than the sequence of do+wait called from the outside?

Reply:
done.